services/src/main/java/io/grpc/services/BinaryLog.java

```diff
@@ -155,16 +152,21 @@ void logTrailingMetadata(Metadata metadata, boolean isServer, byte[] callId) {
         boolean compressed,
         boolean isServer,
         byte[] callId) {
-      ByteBuffer buf = serialize(marshaller, consumeOnce);
+      InputStream stream = marshaller.stream(consumeOnce);
+      if (!(stream instanceof ByteArrayRetainedInputStream)) {
+        throw new IllegalStateException(
+            "Expected the output of the BinaryLog's special ByteArrayMarshaller");
+      }
+      byte[] bytes = ((ByteArrayRetainedInputStream) stream).getBytes();
```

Comment:
Why not:
```
if (!(marshaller instanceof BYTEARRAY_MARSHALLER)) {
  throw ...;
}
byte[] bytes = (byte[]) consumeOnce;
```

That seems functionally equivalent with less "dancing". Then you don't need the ByteArrayRetainedInputStream goo.

Reply:
Switched to casting